server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        try (MockLog mockLog = MockLog.capture(ExceptionsHelper.class)) {
+            mockLog.addExpectation(
+                new MockLog.PatternSeenEventExpectation(
+                    "action listener not called",
+                    ExceptionsHelper.class.getName(),
+                    Level.ERROR,
+                    ".*Expected listener NoopActionListener to be called at least once, but it was never called\\. Created:(.*|\\s)*"
+                )
+            );
+            final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+                ActionListener.assertAtLeastOnce(ActionListener.noop())
+            );
+            assertBusy(() -> {
+                // Nullify reference so it becomes unreachable
+                listenerRef.set(null);
+                System.gc();
+                mockLog.assertAllExpectationsMatched();
+            });
+        }
+    }
+
+    public void testAssertAtLeastOnceWillInvokeListenerWhenNotResolved() throws Exception {
+        AtomicReference<Exception> listenerCreatedAt = new AtomicReference<>();
+        AtomicReference<ActionListener<?>> notCalledListener = new AtomicReference<>();
+        AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+            new ActionListenerImplementations.AssertAtLeastOnceActionListener<>(ActionListener.noop(), (listener, createdAt) -> {
+                logger.info("NotCalledListener called!");
+                notCalledListener.set(listener);
+                listenerCreatedAt.set(createdAt);
+            })
+        );
+        assertBusy(() -> {
+            // Nullify reference so it becomes unreachable
+            listenerRef.set(null);
+            System.gc();
+            assertNotNull(notCalledListener.get());
+            assertNotNull(listenerCreatedAt.get());
+        });
+    }
+
+    public void testAssertAtLeastOnceWillNotLogWhenResolvedOrFailed() throws Exception {
+        final ReachabilityChecker reachabilityChecker = new ReachabilityChecker();
+        final AtomicBoolean notCalledListenerCalled = new AtomicBoolean();
+        AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+            reachabilityChecker.register(
+                new ActionListenerImplementations.AssertAtLeastOnceActionListener<>(
+                    ActionListener.noop(),
+                    (listener, createdAt) -> notCalledListenerCalled.set(true)
+                )
+            )
+        );
+        // Call onResponse and/or onFailure at least once
+        IntStream.range(0, 1 + randomInt(3)).forEach(i -> {
+            if (randomBoolean()) {
+                listenerRef.get().onResponse("succeeded");
+            } else {
+                listenerRef.get().onFailure(new RuntimeException("Failed"));
+            }
+        });
```

Comment:
I do not think it is expected that a listener is called more than once

Reply:
OK that's good to know. I considered that it might be something that happens, due to the presence of e.g. `notifyOnce` etc. Happy to remove the loop if we think it's noise?

Reply:
removed! thanks